app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php

```diff
@@ -211,6 +227,21 @@ public function execute(\Magento\Framework\Event\Observer $observer)
                 continue;
             }
 
+            // Note: we acquire the lock here instead of above, as it should be taken by standalone (child) process,
+            // not by the parent process.
+            if (!$this->lockManager->acquireLock(self::LOCK_PREFIX . $groupId, self::LOCK_TIMEOUT)) {
```

Comment:
Regarding the groups vs jobs discussion. Groups are just logical grouping for the user. There is no guarantee that different groups would not have dependent jobs. So locking mechanism should be implemented per job.

Dependent jobs is a good idea for the future cron evolution.

Reply:
Do we have any examples where one cron job has dependency to another one? Also currently they could be ran in parallel, so we already have the issue. Not sure we need to introduce fix for it as part of this PR.

Reply:
I previously commented that I would prefer cron groups being removed, but let's wait comments from Core team.

At this moment Groups are logical grouping, which would itself be ideal for this use. I can see scenario where example product import related jobs could be put into the same group, so that they do not run parallel, causing less stress on indexing and avoiding possible deadlocks.

Reply:
For now, groups are used for grouping settings (cron run frequency) and it's not guaranteed that jobs from different groups can be run in parallel. As I mentioned before, I agree that it would be good to implement cron job dependencies, so some jobs could run in parallel. For now, jobs should run consequently, independently on the group.

Note: I represent core team, and discussed this question internally.

Reply:
@buskamuza I am not sure how to proceed with this PR after your comment. We can move locking mechanism outside "foreach" loop and have single global lock for cron runner, at least as a stop gap measure.

Yet I would like to see Magento being able to run at least some tasks parallel. Cron groups currently implement a toggle "run in a separate process". Would it be safe to assume those groups can run parallel to main process, otherwise what would be the purpose of this setting?

Reply:
As cron already runs per groups, let's just leave your implementation of locking in scope of this PR.
In future, we may reconsider it and reimplement it based on job dependencies.

Reply:
Just for clarification: no further actions are needed here. Just review other comments in the PR. Thanks.

Reply:
After internal discussion with a development team that faced some issue with 3rd party module, we thought that maybe bit more strict path could be taken here. Implementation could be changed to use single global lock for all groups **except** for those that have "*run in a separate process*" enabled.

Thoughts on this? I would say it's safer way to go.